### PR TITLE
[caffe2][nomnigraph] Add fuseNNPACKConvRelu

### PR DESCRIPTION
**Pull request by bwasti**
simple NNPACK version of convrelu fusion (which doesn't need a new operator)

### CODE REVIEW

File: caffe2/opt/mobile.cc [hunk 1]
```diff
@@ -72,5 +74,40 @@ caffe2::NetDef addNNPACK(caffe2::NetDef net, bool low_memory) {
...
+    auto* op = static_cast<caffe2::OperatorDef*>(annotation->getSaved());
+    if (op->engine() != "NNPACK") { continue; }
+
+    auto relu_outputs = repr::nn::getOutputs(relu_node);
+    if (relu_outputs.size() != 1) { continue; }
+    auto relu_output = relu_outputs.front();
+
+    nn.dataFlow.createEdge(conv_node, relu_output);
+    nn.dataFlow.deleteNode(relu_node);
+    nn.dataFlow.deleteNode(conv_output);
```
Comment (Maratyszcza) [inline caffe2/opt/mobile.cc]: What if relu was in-place (i.e. `conv_output` is `relu_output`)?

File: caffe2/opt/mobile.cc [hunk 2]
```diff
@@ -72,5 +74,40 @@ caffe2::NetDef addNNPACK(caffe2::NetDef net, bool low_memory) {
...
+
+    auto consumers = repr::nn::getConsumers(conv_output);
+    if (consumers.size() != 1) { continue; }
+    if (!repr::nn::is<repr::Relu>(consumers.front())) { continue; }
+    auto relu_node = consumers.front();
+
+    auto annotation = conv->getMutableAnnotation();
+    if (!annotation) { continue; }
+    auto* op = static_cast<caffe2::OperatorDef*>(annotation->getSaved());
+    if (op->engine() != "NNPACK") { continue; }
```
Comment (Maratyszcza) [inline caffe2/opt/mobile.cc]: Right now, NNPACK's Conv+ReLU is efficient only when fast convolution is used, i.e.
```
engine == "WINOGRAD" || engine == "WINOGRAD_FP16" || engine == "FT8x8" || engine == "FT16x16" ||
max(stride) == 1 && min(kernel) >= 2

File: caffe2/opt/mobile.cc (hunk 1 above)
Comment (bwasti) [reply caffe2/opt/mobile.cc]: doesn't matter -- everything is SSA and the "inplace" is recovered by the caffe2 naming convention.

however this reminded me of the case in which neither is inplace but the result would be inplace.  I'll change the logic

File: caffe2/opt/mobile.cc [hunk 3]
```diff
@@ -72,5 +75,60 @@ caffe2::NetDef addNNPACK(caffe2::NetDef net, bool low_memory) {
...
+    if (!repr::nn::is<repr::Relu>(consumers.front())) { continue; }
+    auto relu_node = consumers.front();
+
+    auto annotation = conv->getMutableAnnotation();
+    if (!annotation) { continue; }
+    auto* op = static_cast<caffe2::OperatorDef*>(annotation->getSaved());
+
+    // We only want to fuse for fast NNPACK convs
+    if (op->engine() != "NNPACK") { continue; }
+    caffe2::string algo = "";
```
Comment (Maratyszcza) [inline caffe2/opt/mobile.cc]: You should set algorithm to "AUTO" by default, and check `algo == ""`. The relevant code is:

https://github.com/pytorch/pytorch/blob/master/caffe2/contrib/nnpack/nnpack_ops.cc#L38-L53

File: caffe2/opt/mobile.cc [hunk 4]
```diff
@@ -72,5 +75,60 @@ caffe2::NetDef addNNPACK(caffe2::NetDef net, bool low_memory) {
...
+    for (auto arg : op->arg()) {
+      if (arg.name() == "algo") {
+        algo = arg.s();
+      }
+    }
+    if (algo == "") {
+      bool cont = false;
```
Comment (Maratyszcza) [inline caffe2/opt/mobile.cc]: I think goto is a better solution to nexted `continue` problem. Alternatively, define a small function `inline bool isNNPACKConvReluEfficient(...)`

**Review by Maratyszcza — APPROVED** [review]
LGTM, please fix issues in the comments before merging

**Pull request merged**